### PR TITLE
Add failed status when uploading supporting documents in web banking

### PR DESCRIPTION
**Pull request by sandrine-ds**
Related PR in lake repo:
https://github.com/swan-io/lake/pull/47

Goal of this PR: small refacto in upload document in order to be consistent with the updated logic in Lake UploadArea.

### CODE REVIEW

**Comment by Epimodev** [comment]
ℹ️  Lake v1.9.0 is just published, so you should be able to complete this PR by just upgrading `@swan-io/lake` and `@swan-io/shared-business` to `v1.9.0`

File: clients/banking/src/components/SupportingDocumentsForm.tsx [hunk 1]
```diff
@@ -255,15 +255,22 @@ export const SupportingDocumentsForm = forwardRef<SupportingDocumentsFormRef, Pr
             id: supportingDocumentId,
             name: file.name,
             progress: 0,
-            finished: false,
+            status: "uploading",
           };
 
-          const finalState: LocalDocument = {
+          const finalStateFinished: LocalDocument = {
```
Comment (zoontek) [inline clients/banking/src/components/SupportingDocumentsForm.tsx:261]: It's a good practice to declare variables the closer from where they are used.

File: clients/banking/src/components/SupportingDocumentsForm.tsx [hunk 2]
```diff
@@ -255,15 +255,22 @@ export const SupportingDocumentsForm = forwardRef<SupportingDocumentsFormRef, Pr
...
             purpose,
             id: supportingDocumentId,
             name: file.name,
-            progress: 100,
-            finished: true,
+            status: "finished",
+          };
+
+          const finalStateError: LocalDocument = {
```
Comment (zoontek) [inline clients/banking/src/components/SupportingDocumentsForm.tsx:268]: This variable is used once

File: clients/banking/src/components/SupportingDocumentsForm.tsx [hunk 3]
```diff
@@ -276,7 +283,7 @@ export const SupportingDocumentsForm = forwardRef<SupportingDocumentsFormRef, Pr
               const index = clone.findIndex(item => item.id === supportingDocumentId);
               const item = clone[index];
 
-              if (isNotNullish(item)) {
+              if (isNotNullish(item) && item.status === "uploading") {
```
Comment (zoontek) [inline clients/banking/src/components/SupportingDocumentsForm.tsx:286]: You can use nullish coalescing operator instead.

File: clients/banking/src/components/SupportingDocumentsForm.tsx [hunk 4]
```diff
@@ -290,15 +297,20 @@ export const SupportingDocumentsForm = forwardRef<SupportingDocumentsFormRef, Pr
           };
 
           xhr.onload = () => {
+            const uploadFailed = xhr.status !== 200 && xhr.status !== 204;
             setLocalDocuments(prevState => {
               const clone = [...prevState];
               const index = clone.findIndex(item => item.id === supportingDocumentId);
               const item = clone[index];
 
               if (isNotNullish(item)) {
```
Comment (zoontek) [inline clients/banking/src/components/SupportingDocumentsForm.tsx:306]: Prefer the early return pattern for readability

Comment (zoontek) [reply clients/banking/src/components/SupportingDocumentsForm.tsx:306]: Applying all this feedbacks results in something like:

```ts
getAwsUrl(file, purpose)
  .then(({ supportingDocumentId, upload: { url, fields } }) => {
    const xhr = new XMLHttpRequest();
    xhr.open("POST", url, true);

    const initialState: LocalDocument = {
      purpose,
      id: supportingDocumentId,
      name: file.name,
      progress: 0,
      status: "uploading",
    };

    setLocalDocuments(prevState => [...prevState, initialState]);

    xhr.upload.onprogress = event => {
      const progress = (event.loaded / event.total) * 100;

      setLocalDocuments(prevState => {
        const clone = [...prevState];
        const index = clone.findIndex(item => item.id === supportingDocumentId);
        const item = clone[index];

        if (item?.status === "uploading") {
          clone[index] = { ...item, progress };
        } else {
          clone.push(initialState);
        }

        return clone;
      });
    };

    xhr.onload = () => {
      const uploadFailed = xhr.status !== 200 && xhr.status !== 204;

      const finishedState: LocalDocument = {
        purpose,
        id: supportingDocumentId,
        name: file.name,
        status: "finished",
      };

      setLocalDocuments(prevState => {
        const clone = [...prevState];
        const index = clone.findIndex(item => item.id === supportingDocumentId);
        const item = clone[index];

        if (isNullish(item)) {
          clone.push(finishedState);
          return clone;
        }

        if (!uploadFailed) {
          clone[index] = finishedState;
          return clone;
        }

        clone[index] = {
          purpose,
          id: supportingDocumentId,
          name: file.name,
          error: t("error.generic"),
          status: "failed",
        };

        return clone;
      });
    };

    const formData = new FormData();
    fields.forEach(({ key, value }) => formData.append(key, value));
    formData.append("file", file);

    xhr.send(formData);
  })
  .catch(() => {
    showToast({ variant: "error", title: t("error.generic") });
  });
```

**Review by zoontek — APPROVED** [review]

**Pull request merged**